models/issue_dependency.go

```diff
@@ -0,0 +1,172 @@
+	DependencyID int64          `xorm:"UNIQUE(issue_dependency) NOT NULL"`
+	CreatedUnix  util.TimeStamp `xorm:"INDEX created"`
+	UpdatedUnix  util.TimeStamp `xorm:"updated"`
+}
+
+// DependencyType Defines Dependency Type Constants
+type DependencyType int
+
+// Define Dependency Types
+const (
+	DependencyTypeBlockedBy DependencyType = iota
+	DependencyTypeBlocking
+)
+
+// CreateIssueDependency creates a new dependency for an issue
+func CreateIssueDependency(user *User, issue, dep *Issue) (err error) {
+	sess := x.NewSession()
+	defer sess.Close()
+	if err = sess.Begin(); err != nil {
+		return err
+	}
+
+	// Check if it aleready exists
+	err = issueDepExists(sess, issue.ID, dep.ID)
+	if err != nil {
+		return
+	}
+
+	// No error means the dependency doesn't exist nor is circular, so we can proceed to create a new one.
+	newIssueDependency := &IssueDependency{
+		UserID:       user.ID,
+		IssueID:      issue.ID,
+		DependencyID: dep.ID,
+	}
+
+	if _, err := sess.Insert(newIssueDependency); err != nil {
+		return err
+	}
+
+	// Add comment referencing the new dependency
+	if _, err = createIssueDependencyComment(sess, user, issue, dep, true); err != nil {
+		return
+	}
+
+	// Create a new comment for the dependent issue
+	if _, err = createIssueDependencyComment(sess, user, dep, issue, true); err != nil {
+		return
+	}
+
+	return sess.Commit()
+}
+
+// RemoveIssueDependency removes a dependency from an issue
+func RemoveIssueDependency(user *User, issue *Issue, dep *Issue, depType DependencyType) (err error) {
+	sess := x.NewSession()
+	defer sess.Close()
+	if err = sess.Begin(); err != nil {
+		return err
+	}
+
+	// Check if it exists
+	var exists bool
+	switch depType {
+	case DependencyTypeBlockedBy:
+		err = issueDepExists(sess, issue.ID, dep.ID)
+	case DependencyTypeBlocking:
+		err = issueDepExists(sess, dep.ID, issue.ID)
+	default:
+		return ErrUnknownDependencyType{depType}
+	}
+	if err != nil {
+		if IsErrDependencyExists(err) {
+			exists = true
+		} else {
+			return err
+		}
+	}
+
+	// If it exists, remove it
+	if !exists {
+		return nil
+	}
+
+	var issueDepToDelete IssueDependency
+
+	switch depType {
+	case DependencyTypeBlockedBy:
+		issueDepToDelete = IssueDependency{IssueID: issue.ID, DependencyID: dep.ID}
+	case DependencyTypeBlocking:
+		issueDepToDelete = IssueDependency{IssueID: dep.ID, DependencyID: issue.ID}
+	default:
+		return ErrUnknownDependencyType{depType}
+	}
+
+	if _, err := sess.Delete(&issueDepToDelete); err != nil {
+		return err
+	}
+
+	// Add comment referencing the removed dependency
+	if _, err = createIssueDependencyComment(sess, user, issue, dep, false); err != nil {
+		return err
+	}
+
+	// Create a new comment for the dependent issue
+	if _, err = createIssueDependencyComment(sess, user, dep, issue, false); err != nil {
+		return err
+	}
+	return sess.Commit()
+}
+
+// Check if the dependency already exists
+func issueDepExists(e Engine, issueID int64, depID int64) (err error) {
```

Comment:
Why returning error instead of `true/false`? You should return both.

Reply:
Because it returns an error `DependencyDoesNotExists` when the dependency does not exist.

I've had that function returning a boolean...

Reply:
Is it rly an error? Not existence of something is error (when you are just asking if it exists or not)? I don't see any reason why answering yes/no question should be error. `error` should be used only for unexpected behaviour.

Maybe you wanted to know why it returns false. I see it is used only in 2 functions, and only one need both checks (regular and "circular"), another one ignores "circular" dep. IMO it should check only one thing and if you need to check both options, call it twice. Error (message) should be constructed (e.g. `ErrDependencyExists` and `ErrCircularDependency`) in caller func.

Reply:
You mean two seperate functions, one which checks if `DepenendcyExists` and one which checks if `IsCircularDependency`?

Reply:
I've redone `issueDepExists` in two functions (which return booleans!), as I stated above. It seems logic to the way you explain it, I don't know why I did it the way it was done back then.

I guess this is what happens when you don't look at all parts of your code for quite some time :)